Mathlib/Computability/DFA.lean

```diff
@@ -163,7 +174,84 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [Set.mem_singleton_iff] at ha' hc'
   substs ha' hc'
   have h := M.evalFrom_of_pow hb hb'
-  rwa [mem_accepts, evalFrom_of_append, evalFrom_of_append, h, hc]
+  rwa [mem_accepts', evalFrom_of_append, evalFrom_of_append, h, hc]
 #align DFA.pumping_lemma DFA.pumping_lemma
 
+section Comap
+
+variable {α' : Type*} (f : α' → α)
+
+/-- Transforms the alphabet of a DFA. -/
+@[simps]
+def comap : DFA α' σ where
+  step s a := M.step s (f a)
+  start := M.start
+  accept := M.accept
+
+@[simp]
+theorem comap_evalFrom (s : σ) (x : List α') :
+    (M.comap f).evalFrom s x = M.evalFrom s (x.map f) := by
+  induction x using List.list_reverse_induction with
+  | base => simp
+  | ind x a ih => simp [ih]
+
+@[simp]
+theorem comap_eval (x : List α') : (M.comap f).eval x = M.eval (x.map f) := by simp [eval]
+
+@[simp]
+theorem comap_accepts : (M.comap f).accepts = List.map f ⁻¹' M.accepts := by
+  ext x
+  conv =>
+    rhs
+    rw [Set.mem_preimage, mem_accepts]
+  simp [mem_accepts]
+
+end Comap
+
+section Reindex
+
+variable {σ' : Type*} (g : σ ≃ σ')
+
+/-- Lifts an equivalence on states to an equivalence on DFAs. -/
+def reindex : DFA α σ ≃ DFA α σ' where
+  toFun M := {
+    step := fun s a => g (M.step (g.symm s) a)
+    start := g M.start
+    accept := g.symm ⁻¹' M.accept
+  }
+  invFun M := {
+    step := fun s a => g.symm (M.step (g s) a)
+    start := g.symm M.start
+    accept := g ⁻¹' M.accept
+  }
+  left_inv M := by simp
+  right_inv M := by simp
+
+@[simp]
+theorem reindex_step (s : σ') (a : α) : (reindex g M).step s a = g (M.step (g.symm s) a) := by
+  simp [reindex]
+
+@[simp]
+theorem reindex_start : (reindex g M).start = g M.start := by simp [reindex]
+
+@[simp]
+theorem reindex_accept : (reindex g M).accept = g.symm ⁻¹' M.accept := by simp [reindex]
```

Comment:
can `simps` generate these?

Reply:
Huh, I thought `simps` doesn't work on `Equiv`, but it turns out it does!

Reply:
Nevermind. The `@[simps]` attribute generates `reindex_symm_step` etc. which conflict with the manually defined `reindex_symm`.

```
/- The `simpNF` linter reports:
SOME SIMP LEMMAS ARE NOT IN SIMP-NORMAL FORM.
see note [simp-normal form] for tips how to debug this.
https://leanprover-community.github.io/mathlib_docs/notes.html#simp-normal%20form -/
-- Mathlib.Computability.DFA
Error: ./././Mathlib/Computability/DFA.lean:219:3: error: DFA.reindex_symm_apply_step.{u_1, u, v} simp can prove this:
  by simp only [@DFA.reindex_symm, @DFA.reindex_apply_step, @Equiv.symm_symm]
One of the lemmas above could be a duplicate.
If that's not the case try reordering lemmas or adding @[priority].

Error: ./././Mathlib/Computability/DFA.lean:219:3: error: DFA.reindex_symm_apply_accept.{u_1, u, v} simp can prove this:
  by simp only [@DFA.reindex_symm, @DFA.reindex_apply_accept, @Equiv.symm_symm]
One of the lemmas above could be a duplicate.
If that's not the case try reordering lemmas or adding @[priority].
```

Reply:
I found out there's `@[simps foo bar baz]` syntax that lets me specify the projections explicitly.